object_store_factory/src/aws.rs

```diff
@@ -49,38 +77,116 @@ impl S3Config {
+                .unwrap_or(true),
+            skip_signature: map
+                .remove("format.skip_signature")
+                .map(|s| s != "false")
+                .unwrap_or(true),
         })
     }
-}
 
-pub fn build_amazon_s3_from_config(
-    config: &S3Config,
-) -> Result<Arc<dyn ObjectStore>, object_store::Error> {
-    let mut builder = AmazonS3Builder::new()
-        .with_region(config.region.clone().unwrap_or_default())
-        .with_bucket_name(config.bucket.clone())
-        .with_allow_http(true);
+    pub fn to_hashmap(&self) -> HashMap<String, String> {
+        let mut map = HashMap::new();
+        if let Some(region) = &self.region {
+            map.insert(
+                AmazonS3ConfigKey::Region.as_ref().to_string(),
+                region.clone(),
+            );
+        }
+        if let Some(access_key_id) = &self.access_key_id {
+            map.insert(
+                AmazonS3ConfigKey::AccessKeyId.as_ref().to_string(),
+                access_key_id.clone(),
+            );
+        }
+        if let Some(secret_access_key) = &self.secret_access_key {
+            map.insert(
+                AmazonS3ConfigKey::SecretAccessKey.as_ref().to_string(),
+                secret_access_key.clone(),
+            );
+        }
+        if let Some(session_token) = &self.session_token {
+            map.insert(
+                AmazonS3ConfigKey::Token.as_ref().to_string(),
+                session_token.clone(),
+            );
+        }
+        if let Some(endpoint) = &self.endpoint {
+            map.insert(
+                AmazonS3ConfigKey::Endpoint.as_ref().to_string(),
+                endpoint.clone(),
+            );
+        }
+        map.insert(
+            AmazonS3ConfigKey::Bucket.as_ref().to_string(),
+            self.bucket.clone(),
+        );
+        if let Some(prefix) = &self.prefix {
+            map.insert("prefix".to_string(), prefix.clone());
+        }
+        map.insert(
+            AmazonS3ConfigKey::Client(ClientConfigKey::AllowHttp)
+                .as_ref()
+                .to_string(),
+            self.allow_http.to_string(),
+        );
+        map.insert(
+            AmazonS3ConfigKey::SkipSignature.as_ref().to_string(),
+            self.skip_signature.to_string(),
+        );
+        map
+    }
 
-    if let Some(endpoint) = &config.endpoint {
-        builder = builder.with_endpoint(endpoint.clone());
+    pub fn bucket_to_url(&self) -> String {
+        format!("s3://{}", &self.bucket)
     }
 
-    if let (Some(access_key_id), Some(secret_access_key)) =
-        (&config.access_key_id, &config.secret_access_key)
-    {
-        builder = builder
-            .with_access_key_id(access_key_id.clone())
-            .with_secret_access_key(secret_access_key.clone());
+    pub fn build_amazon_s3(&self) -> Result<Arc<dyn ObjectStore>, object_store::Error> {
+        assert!(self.allow_http, "allow_http must be true for Amazon S3");
```

Comment:
This should be the case only for testing on e.g. MinIO, otherwise it should be disallowed (no point in `allow_http` field if this assert is here).

Reply:
I forgot to remove this. I wasn't sure if the other code at any point calls this with `allow_http == false`, so I added this to catch. I will remove

Reply:
Ah makes sense, thanks.